Foundation/src/ProcessRunner.cpp

```diff
@@ -0,0 +1,214 @@
+//
+// ProcessRunner.cpp
+//
+// Library: Foundation
+// Package: Processes
+// Module:  ProcessRunner
+//
+// Copyright (c) 2023, Applied Informatics Software Engineering GmbH.
+// Aleph ONE Software Engineering d.o.o.,
+// and Contributors.
+//
+// SPDX-License-Identifier:    BSL-1.0
+//
+
+
+#include "Poco/ProcessRunner.h"
+#include "Poco/PIDFile.h"
+#include "Poco/FileStream.h"
+#include "Poco/AutoPtr.h"
+#include "Poco/File.h"
+#include "Poco/Path.h"
+#include "Poco/String.h"
+#include <fstream>
+
+
+using Poco::Thread;
+using Poco::Process;
+using Poco::ProcessHandle;
+using Poco::FileInputStream;
+using Poco::AutoPtr;
+using Poco::File;
+using Poco::Path;
+using Poco::Stopwatch;
+
+
+namespace Poco {
+
+
+ProcessRunner::ProcessRunner(const std::string& cmd,
+		const Args& args,
+		const std::string& pidFile,
+		int options,
+		int timeout,
+		bool startProcess,
+		const Args& pidArgFmt): _cmd(cmd),
+			_args(args),
+			_pid(INVALID_PID),
+			_pidFile(pidFile),
+			_options(options),
+			_timeout(timeout),
+			_pPH(nullptr),
+			_started(false),
+			_rc(RESULT_UNKNOWN)
+{
+	if (!File(_cmd).exists())
+		throw Poco::NotFoundException(_cmd);
+
+	if (_pidFile.empty() && !_args.empty() && !pidArgFmt.empty())
+	{
+		// try to detect PID file name from arguments;
+		// we follow Poco convention
+		// (-pfile or --pidfile=file)
+		// TODO: windows
+		for (const auto& fmt : pidArgFmt)
+		{
+			for (const auto& arg : _args)
+			{
+				std::string a = Poco::trim(arg);
+				std::size_t pos = a.find(fmt);
+				if (pos == 0)
+				{
+					_pidFile = a.substr(fmt.length());
+					PIDFile::getFileName(_pidFile);
+					break;
+				}
+			}
+		}
+	}
+	if (startProcess) start();
+}
+
+
+ProcessRunner::~ProcessRunner()
+{
+	stop();
+	_t.join();
+}
+
+
+std::string ProcessRunner::cmdLine() const
+{
+	std::string cmdL = _cmd + ' ';
+	auto it = _args.begin();
+	auto end = _args.end();
+	for (; it != end;)
+	{
+		cmdL.append(*it);
+		if (++it == end) break;
+		cmdL.append(1, ' ');
+	}
+	return cmdL;
+}
+
+
+void ProcessRunner::run()
+{
+	ProcessHandle* pPH = nullptr;
+
+	try
+	{
+		_pPH.store(new ProcessHandle(Process::launch(_cmd, _args, _options)));
+		pPH = _pPH;
+		_pid = pPH->id();
+		_rc = pPH->wait();
+	}
+	catch (...)
+	{
+	}
+
+	_pid = INVALID_PID;
+	_pPH.store(nullptr);
+
+	delete pPH;
+}
+
+
+void ProcessRunner::stop()
+{
+	if (_started)
+	{
+		// TODO:
```

Comment:
Ignore this for now, I was mistaken. I may make sense to implement `signal` on windows, but I'm not sure. Remove the coment for release.